Refactor Poseidon example into its own machine

georgwiese opened the pull request:
Refactors the Poseidon example so that it lives in a separate block machine, so that @gballet can copy-paste it to use in the RISC-V VM.

Two notes on that:
- ~The `degree` field will need to changed so that it's the same as all other machines in the RISC-V VM~
- Since the `poseidon_permutation` function takes 3 inputs and returns 1 output, there will need to be an additional assignment register (which we can undo once #480 is implemented)

leonardoalt commented:
I think now you can also just remove the degree completely and it gets copied from Main? At least in asm? @Schaeff

leonardoalt commented on test_data/asm/poseidon_bn254.asm:
```diff
@@ -1,11 +1,15 @@
+machine Poseidon(LASTBLOCK, function_id) {
+
+    degree 1024;
```
If you remove the degree it should just be either copied from Main or assigned 1024 automatically by the linker I think

Schaeff replied:
https://powdr-labs.github.io/powdr/linker/index.html

georgwiese commented:
Nice! I removed the degree. I also renamed the machine to `PoseidonBN254` and added a clarifying comment about it only working with the Halo2 backend.

Schaeff reviewed (commented):
LGTM! Just one thing: can we run the Poseidon test on CI? Maybe just adding it to `compiler/tests/asm.rs`?

georgwiese commented:
It is tested in `halo2/src/mock_prover.rs`!
